### PR TITLE
k8s: allow setting multiple k8s API server addresses

### LINKED ISSUE #19038: k8sServiceHost multiple IPs/Hosts
## Proposal / RFE

**Is your feature request related to a problem?**
no, it's just a inconvenience on bare-metal

**Describe the solution you'd like**
currently we run multiple clusters with cilium and use the kube-proxy'less mode and one thing is always not that good in our bare metal clusters is specifiyng the `k8sServiceHost`. This is especially a problem when a node needs to get rebooted.
We would prefer if it would be possible to  have that load balanced or retried with multiple addresses.

Something like:
```markdown
k8sMasterServices:
  - host: 10.10.30.11
     port: 6443 # somehow github formats that incorrectly?
  - host: 10.10.30.12
     port: 6443
  - host: 10.10.30.13
     port: 6443
```

this would make it much more easy to use the kube-proxy less mode when running on bare-metal since an external load balancer would not be needed, especially when running with something like k3s where the master's do load balancer internally.

(btw. at the moment we use kube-vip for that, but preferably we would drop it, once k8s is up and running you can use metallb and announce the kube-api masters by using a endpoint and svc and feed that to metallb)

### PR DESCRIPTION
**Pull request by fristonio**
See commit message for detailed description

Introduce a new command line parameter(`--k8s-api-server-urls`) and helm option(`k8s.apiServerURLs`) to specify multiple k8s API server addresses for the client to use.

Fixes: #19038 

```release-note
Allow setting of multiple k8s API server URLs for client
```

### CODE REVIEW

**Comment by brb** [comment]
@fristonio Welcome back! :sweat_smile: Would it be possible to set the new param from Helm? Currently, we use this hack to pass the API server endpoint addr - https://github.com/cilium/cilium/blob/master/install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml#L193.

**Comment by fristonio** [comment]
Hey @brb 
Its good to be back. 😄 
Yeah, I have some changes locally for the helm option that I need to test. Will update the PR soon.

File: operator/flags.go
Comment (sayboras) [inline operator/flags.go:238]: what do think about marking this one deprecated in favor of new flags ? Also a short description in this docs https://docs.cilium.io/en/stable/operations/upgrade/#admin-upgrade will be nice.

**Review by sayboras — COMMENTED** [review]
:wave: super happy to see you around :medal_military: 

One small comment as per below, the rest looks reasonable to me.

File: operator/flags.go
Comment (fristonio) [reply operator/flags.go:238]: Yeah, that makes sense to me.
The flag `--k8s-api-server` is not exposed as an helm option anyways so we can easily deprecate it in favor of the new flag.
Will update the flag and the docs.

File: pkg/k8s/client.go
Comment (qmonnet) [inline pkg/k8s/client.go]:
```suggestion
				log.WithField("address", config.APIServerURL).Info("Rotating Kubernetes API server URL for connection")
```

File: pkg/k8s/config.go
Comment (qmonnet) [inline pkg/k8s/config.go]:
```suggestion
// is not responding (failing heartbeat check).
```

File: pkg/k8s/config.go
Comment (qmonnet) [inline pkg/k8s/config.go]: It's an `Error()` already
```suggestion
				log.WithError(err).Error("Failed to parse APIServer URL")
```

File: pkg/k8s/config.go
Comment (qmonnet) [inline pkg/k8s/config.go]: Same, let's avoid “Error” in a warning?

File: pkg/option/config.go
Comment (qmonnet) [inline pkg/option/config.go]: typo: Kuberentes

**Review by qmonnet — CHANGES_REQUESTED** [review]
Looks good, with a few nits - see below (plus a typo on `--k8s-api-erver-addresses` in your first commit log).

One question: I'm wondering if it would make sense to have “URLs” instead of “Addresses” in the flag (and Helm value) name? Reading “addresses” I'd first think of IP addresses, maybe URLs would make it more explicit?

Awesome to have you contributing again! :heart:

**Comment by fristonio** [comment]
Hey @qmonnet! 👋 
Yeah, changing the CLI flag to use `urls` instead of `addresses` makes sense to me.
I have updated the PR and addressed the changes you requested.

File: daemon/cmd/daemon_main.go
Comment (tommyp1ckles) [inline daemon/cmd/daemon_main.go:1210]: Would it be worth mentioning the precedence order in the flag help? Or perhaps a warning if you've set both.

File: operator/main.go
Comment (tommyp1ckles) [inline operator/main.go:143]: Same question as above about the precedence behaviour.

File: pkg/k8s/config.go
Comment (tommyp1ckles) [inline pkg/k8s/config.go]: Just a thought, would it make more sense to parse these once and store them inside the config object as URL objects? Might be able to simplify and remove the need for unreachable code.

**Review by qmonnet — APPROVED** [review]
Good points raised by Tom, but I'm good for my own review. Thanks again! :)

File: daemon/cmd/daemon_main.go
Comment (fristonio) [reply daemon/cmd/daemon_main.go:1210]: Given the old flag is deprecated instead of mentioning precedence order in flag help we can log a warning if the user sets both the options as you mentioned.

File: pkg/k8s/config.go
Comment (fristonio) [reply pkg/k8s/config.go]: Yeah, that makes sense.

**Review by sayboras — APPROVED** [review]
LGTM :100:

**Review by tommyp1ckles — APPROVED** [review]
lgtm, thanks!

File: install/kubernetes/cilium/values.yaml
Comment (joestringer) [inline install/kubernetes/cilium/values.yaml:1178]: After #20066 was merged, now these changes need to be made to `install/kubernetes/cilium/values.yaml.tmpl` instead, then run `make -C install/kubernetes` to update this file.

**Comment by joestringer** [comment]
Looks like this just needs rebase, address any CI failures, plus operator / cli codeowner review. If there's something particular you'd like me to look at, then let me know - otherwise I'll let the other core reviewers follow up.

File: pkg/k8s/config.go
Comment (aanm) [inline pkg/k8s/config.go:121]: These fields are being accessed without being protected by their mutex

File: pkg/k8s/config.go
Comment (aanm) [inline pkg/k8s/config.go:183]: Can you leave a comment stating what this is doing?

File: pkg/option/config.go
Comment (aanm) [inline pkg/option/config.go:3719]: in the `daemon/flags.go` we should be marking the flag as deprecated.

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by christarazi** [comment]
Moving to draft, please mark it ready for review when you are ready to pick it up.

File: pkg/k8s/config.go
Comment (squeed) [inline pkg/k8s/config.go:27]: Can this member be dropped?

**Pull request closed without merging**
